File: worker/retry_test.go
package main

import (
	"context"
	"testing"
)

func TestRetryFailedJobs(t *testing.T) {
	q := testLocalQueue(t)
	a, _ := q.submit([]string{"false"})
	q.claim()
	code := 1
	if err := q.Finish(context.Background(), a.ID, Result{ExitCode: &code, Error: "failed"}); err != nil {
		t.Fatal(err)
	}
	b, _ := q.submit([]string{"true"})
	if _, err := q.retry(b.ID); err == nil {
		t.Fatal("retried queued job")
	}
	if _, err := q.retry("local-999"); err == nil {
		t.Fatal("retried missing job")
	}
	result, err := q.retry(a.ID)
	if err != nil || result.Retried != 1 {
		t.Fatalf("retry: %+v %v", result, err)
	}
	reset, _ := q.get(a.ID)
	if reset.Status != "queued" || reset.StartedAt != nil || reset.FinishedAt != nil || reset.ExitCode != nil || reset.Error != "" || reset.WorkerID != "" || reset.OutputPath != "" || reset.CancelRequested != 0 {
		t.Fatalf("reset: %+v", reset)
	}
	next, _ := q.claim()
	if next.ID != b.ID {
		t.Fatal("retry jumped ahead of queued work")
	}
	q.Finish(context.Background(), b.ID, Result{ExitCode: &code, Error: "failed"})
	result, err = q.retry("")
	if err != nil || result.Retried != 1 {
		t.Fatalf("retry all: %+v %v", result, err)
	}
	result, err = q.retry("")
	if err != nil || result.Retried != 0 {
		t.Fatalf("empty retry all: %+v %v", result, err)
	}
	next, _ = q.claim()
	if next.ID != a.ID {
		t.Fatal("retry order changed")
	}
}

File: controller/test/retry.test.ts
import { expect, it } from 'vitest';
import { schedulerStorage } from './storage';

it('requeues failed jobs with the same IDs behind queued work and resets execution fields', async () => {
  const { scheduler: s, api } = schedulerStorage();
  s.register('w', 'host');
  const a = s.submit(['false']);
  s.claim('w');
  s.output(a.id, 'w', '/tmp/old.log');
  s.cancel(a.id);
  s.finish(a.id, 'w', false, 1, 'failed');
  const b = s.submit(['false']);
  s.claim('w');
  s.finish(b.id, 'w', false, 1, 'failed');
  const queued = s.submit(['true']);
  const post = (path: string) => api.request(path, { method: 'POST' });
  expect((await post(`/jobs/${queued.id}/retry`)).status).toBe(409);
  expect((await post('/jobs/missing/retry')).status).toBe(404);
  expect(await (await post(`/jobs/${a.id}/retry`)).json()).toEqual({ retried: 1 });
  expect(s.job(a.id)).toMatchObject({
    id: a.id, status: 'queued', command: ['false'], created_at: a.created_at,
    started_at: null, finished_at: null, worker_id: null, output_path: null,
    exit_code: null, error: null, cancel_requested: 0,
  });
  expect(await (await post('/jobs/retry-all')).json()).toEqual({ retried: 1 });
  expect(await (await post('/jobs/retry-all')).json()).toEqual({ retried: 0 });
  for (const id of [queued.id, a.id, b.id]) {
    expect(s.claim('w')?.id).toBe(id);
    s.finish(id, 'w', true, 0, null);
  }
  expect(s.retry()).toEqual({ retried: 0 });
});
